File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
   
map<char, char> mp = {{'{', '}'}, {'[', ']'}, {'(', ')'}};


bool isValid(string s)
{
  stack<char> st;
  if (s.length() % 2 != 0)
  {
    return false;
  }

  for (int i = 0; i < s.length(); i++)
  {
    if (openParentheses(s[i]))
    {
      st.push(s[i]);
    }
    else
    {
      if (!st.empty() && mp[st.top()] == s[i])
      {
        st.pop();
      }
      else
      {
        return false;
      }
    }
  }
  if (st.empty())
  {
    return true;
  }
  return false;
}
bool openParentheses(char c1)
{
  if (c1 == '{' || c1 == '[' || c1 == '(')
  {
    return true;
  }
  return false;
}
};

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    Map<Character,Character> mp = Map.of(
        ')','(',
        ']','[',
        '}','{'
    );
    
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for(int i=0;i<s.length();i++){
            char current = s.charAt(i);
            if(mp.containsKey(current)){
               if(stack.isEmpty() || stack.pop() != mp.get(current)) {
                    return false;
                }
            }else{
                stack.push(current);
            }
        }
        return stack.isEmpty();
    }
}